packages/frontend-app-api/src/wiring/createSpecializedApp.tsx

```diff
@@ -200,6 +202,15 @@ export function createSpecializedApp(options?: {
   bindRoutes?(context: { bind: CreateAppRouteBinder }): void;
 }): { createRoot(): JSX.Element } {
   const config = options?.config ?? new ConfigReader({}, 'empty-config');
+  const featureLoaders = options?.features?.filter(
+    isInternalFrontendFeatureLoader,
+  );
+  if (featureLoaders && featureLoaders.length > 0) {
+    throw new Error(
+      `FrontendFeatureLoader is an async construct. It should be passed to createApp(), not to createSpecializedApp().`,
+    );
+  }
```

Comment:
So this is the kind of issue I was alluding the regarding the new meaning of the `FrontendFeature` type. Tbh, I think we should keep `FrontendFeature` where it is and as it is, and make `createApp` accept `FrontendFeature | CreateAppFeatureLoader | FrontendFeatureLoader` instead. I at least can't think of any reason right now why that wouldn't be enough and I think it's better to avoid this confusion and keep it type safe.

Reply:
So I guess that the `loader` method of the `FrontendFeatureLoader` itself should then return `FrontendFeature | FrontendFeatureLoader` ?
Having a single Type for anything that can be loaded by the feature loader (including nested feature loaders themselves) is quite useful.
The same remark applies to the discovery `getAvailableFeatures()` method, which we expect to possibly return feature loaders as well.

Reply:
Is a change in the types returned by `FrontendFeatureLoader.loader` and `getAvailableFeatures ()` the direction you would recommend ?

Reply:
@Rugvip Please could you provide your thought here, before I update the PR and rebase it on top of https://github.com/backstage/backstage/pull/29058 ?

Reply:
yep, essentially any point in the code that requires a sync logic would use `sync`, while any point that can do async as well would have `sync | async`